File: codex_statusline/domain/render.lua
local M = {}
local version = require("codex_statusline.version")
local common = require("codex_statusline.domain.common")
local session = require("codex_statusline.domain.session")
local pricing = require("codex_statusline.domain.pricing")
local BASELINE_TOKENS = session.BASELINE_TOKENS
local GOAL_STATUSES = session.GOAL_STATUSES
local trim = common.trim
local to_number = common.to_number
local normalized_event_type = session.normalized_event_type
local MODEL_ICONS = { astra = "✦", sol = "☀", terra = "⊕", luna = "☾" }

function M.model_text(model, display)
  if not model or (display ~= "icon_name" and display ~= "icon") then
    return model
  end
  for token in model:lower():gmatch("[a-z0-9]+") do
    local icon = MODEL_ICONS[token]
    if icon then
      return display == "icon" and icon or (icon .. " " .. model)
    end
  end
  return model
end

function M.render_layout_key(pane_id, cols, rows, font_size)
  return table.concat({
    tostring(pane_id or "?"),
    tostring(to_number(cols) or "?"),
    tostring(to_number(rows) or "?"),
    tostring(to_number(font_size) or "?"),
  }, "|")
end

local LEGACY_SEGMENT_ORDER = {
  "label",
  "model",
  "reasoning",
  "provider",
  "cwd",
  "git",
  "context",
  "used_tokens",
}

local DEFAULT_SEGMENT_ORDER = {
  "label",
  "model",
  "reasoning",
  "activity",
  "provider",
  "personality",
  "service_tier",
  "cwd",
  "project",
  "git",
  "permissions",
  "approval",
  "context",
  "context_used",
  "context_window",
  "used_tokens",
  "cache_rate",
  "cost",
  "input_tokens",
  "cached_tokens",
  "output_tokens",
  "reasoning_tokens",
  "thread_id",
  "task_progress",
  "codex_version",
  "icon",
}

local DROP_ORDER = {
  "provider",
  "personality",
  "service_tier",
  "codex_version",
  "thread_id",
  "approval",
  "permissions",
  "cached_tokens",
  "reasoning_tokens",
  "input_tokens",
  "output_tokens",
  "context_window",
  "context_used",
  "task_progress",
  "project",
  "cwd",
  "cache_rate",
  "cost",
  "used_tokens",
  "git",
  "reasoning",
  "model",
}

local function compact_trim(value)
  return value:gsub("(%..-)0+$", "%1"):gsub("%.$", "")
end

function M.compact_number(value)
  local number = to_number(value)
  if not number then
    return nil
  end
  local sign = number < 0 and "-" or ""
  local absolute = math.abs(number)
  if absolute < 1000 then
    return sign .. tostring(math.floor(absolute + 0.5))
  end

  local units = {
    { 1000000000000, "T" },
    { 1000000000, "B" },
    { 1000000, "M" },
    { 1000, "K" },
  }
  for _, unit in ipairs(units) do
    if absolute >= unit[1] then
      local scaled = absolute / unit[1]
      local formatted = nil
      if scaled < 10 then
        formatted = string.format("%.2f", scaled)
      elseif scaled < 100 then
        formatted = string.format("%.1f", scaled)
      else
        formatted = string.format("%.0f", scaled)
      end
      local rounded = tonumber(formatted) or scaled
      if rounded >= 1000 then
        for next_index, next_unit in ipairs(units) do
          if next_unit[1] > unit[1] and absolute >= next_unit[1] * 0.9995 then
            local next_scaled = absolute / next_unit[1]
            local next_formatted = next_scaled < 10 and string.format("%.2f", next_scaled)
              or (next_scaled < 100 and string.format("%.1f", next_scaled) or string.format("%.0f", next_scaled))
            return sign .. compact_trim(next_formatted) .. next_unit[2]
          end
        end
      end
      return sign .. compact_trim(formatted) .. unit[2]
    end
  end
  return sign .. tostring(math.floor(absolute + 0.5))
end

function M.context_bar(percent, cells)
  local value = math.max(0, math.min(100, to_number(percent) or 0))
  local width = math.max(1, math.floor(to_number(cells) or 10))
  local filled = math.max(0, math.min(width, math.floor((value / 100) * width + 0.5)))
  return string.rep("█", filled) .. string.rep("░", width - filled)
end

function M.render_layout(columns)
  local cols = math.max(1, math.floor(to_number(columns) or 120))
  if cols < 60 then
    return "tiny"
  end
  if cols < 90 then
    return "narrow"
  end
  if cols < 120 then
    return "medium"
  end
  return "wide"
end

local function list_set(values)
  local result = {}
  for key, value in pairs(type(values) == "table" and values or {}) do
    if type(key) == "number" then
      result[tostring(value)] = true
    elseif value == true then
      result[tostring(key)] = true
    end
  end
  return result
end

local function text_width(value)
  local text = tostring(value or "")
  if type(utf8) == "table" and type(utf8.len) == "function" then
    local ok, length = pcall(utf8.len, text)
    if ok and length then
      return length
    end
  end
  return #text
end

local function plan_width(segments, powerline)
  local width = 0
  for index, segment in ipairs(segments) do
    width = width + text_width(segment.text) + (powerline and 2 or 0)
    if index < #segments then
      width = width + (powerline and 1 or 2)
    end
  end
  return width
end

local function theme_for(options, id)
  local theme = type(options.theme) == "table" and options.theme or {}
  local themes = type(theme.segments) == "table" and theme.segments or {}
  local aliases = {
    reasoning = "thinking",
    context = "tokens",
    context_used = "tokens",
    context_window = "tokens",
    used_tokens = "tokens",
    cache_rate = "tokens",
    cost = "tokens",
    input_tokens = "tokens",
    cached_tokens = "tokens",
    output_tokens = "tokens",
    reasoning_tokens = "tokens",
  }
  local value = themes[id] or themes[aliases[id]] or {}
  return value.bg or theme.bg or "#11151a", value.fg or theme.fg or "#d5dbe3"
end

local function ordered_ids(options)
  local render = type(options.render) == "table" and options.render or {}
  local configured = type(render.segment_order) == "table" and render.segment_order or DEFAULT_SEGMENT_ORDER
  local result = {}
  local seen = {}
  for _, id in ipairs(configured) do
    id = tostring(id)
    if not seen[id] then
      seen[id] = true
      table.insert(result, id)
    end
  end

  local legacy = #configured > 0
  local legacy_set = list_set(LEGACY_SEGMENT_ORDER)
  for _, id in ipairs(configured) do
    if not legacy_set[tostring(id)] then
      legacy = false
      break
    end
  end
  local appendable = legacy and { "activity", "icon" } or DEFAULT_SEGMENT_ORDER
  for _, id in ipairs(appendable) do
    if not seen[id] then
      seen[id] = true
      table.insert(result, id)
    end
  end

  local anchored = {}
  for _, id in ipairs(result) do
    if id ~= "icon" then
      table.insert(anchored, id)
    end
  end
  if seen.icon then
    table.insert(anchored, "icon")
  end
  return anchored
end

local DEFAULT_ACTIVITY_LABELS = {
  plan = "PLAN",
  review = "REVIEW",
  goal_active = "GOAL",
  goal_paused = "GOAL PAUSED",
  goal_blocked = "GOAL BLOCKED",
  goal_usage_limited = "GOAL LIMITED",
  goal_budget_limited = "GOAL BUDGET",
  goal_complete = "GOAL DONE",
}

local function activity_text(options, snapshot)
  local activity = type(snapshot.activity) == "table" and snapshot.activity or {}
  local configured = type(options.activity) == "table" and options.activity.labels or nil
  local labels = type(configured) == "table" and configured or DEFAULT_ACTIVITY_LABELS
  if activity.review == true then
    return labels.review or DEFAULT_ACTIVITY_LABELS.review
  end
  if normalized_event_type(activity.mode) == "plan" then
    return labels.plan or DEFAULT_ACTIVITY_LABELS.plan
  end
  local status = type(activity.goal) == "table" and normalized_event_type(activity.goal.status) or nil
  if status and GOAL_STATUSES[status] then
    local key = "goal_" .. status
    return labels[key] or DEFAULT_ACTIVITY_LABELS[key]
  end
  return nil
end

local function segment_text(id, options, snapshot, layout)
  if id == "label" then
    return snapshot.label or options.label or "CODEX"
  end
  if id == "model" then
    return M.model_text(snapshot.model, type(options.render) == "table" and options.render.model_display or nil)
  end
  if id == "reasoning" then
    return snapshot.reasoning
  end
  if id == "activity" then
    return activity_text(options, snapshot)
  end
  if id == "provider" then
    if layout == "wide" and snapshot.provider then
      return "p:" .. tostring(snapshot.provider)
    end
    return snapshot.provider
  end
  if id == "personality" then
    return snapshot.personality and ("persona:" .. tostring(snapshot.personality)) or nil
  end
  if id == "service_tier" then
    return snapshot.service_tier and ("tier:" .. tostring(snapshot.service_tier)) or nil
  end
  if id == "cwd" then
    return snapshot.cwd
  end
  if id == "project" then
    return snapshot.project
  end
  if id == "git" then
    return snapshot.git
  end
  if id == "permissions" then
    return snapshot.permissions
  end
  if id == "approval" then
    return snapshot.approval
  end
  if id == "thread_id" then
    local value = trim(snapshot.thread_id)
    return value and ("id:" .. value:sub(1, 8)) or nil
  end
  if id == "task_progress" and type(snapshot.task_progress) == "table" then
    local completed = to_number(snapshot.task_progress.completed)
    local total = to_number(snapshot.task_progress.total)
    if completed and total and total > 0 then
      return "Tasks " .. tostring(math.floor(completed)) .. "/" .. tostring(math.floor(total))
    end
    return nil
  end
  if id == "codex_version" then
    local version = trim(snapshot.codex_version)
    return version and ("v" .. version:gsub("^v", "")) or nil
  end
  if id == "icon" then
    return trim(type(options.icon) == "table" and options.icon.text or nil) or ""
  end

  local usage = type(snapshot.usage) == "table" and snapshot.usage or nil
  if id == "context" then
    if not usage then
      return snapshot.waiting
    end
    local percent = to_number(usage.context_remaining_percent)
    if percent ~= nil then
      percent = math.max(0, math.min(100, math.floor(percent + 0.5)))
      if layout == "wide" or layout == "medium" then
        local cells = layout == "wide" and 10 or 8
        return "Ctx " .. M.context_bar(percent, cells) .. " " .. tostring(percent) .. "% left"
      end
      if layout == "narrow" then
        return "Ctx " .. tostring(percent) .. "% left"
      end
      return "Ctx " .. tostring(percent) .. "%"
    end
    local current = M.compact_number(usage.context_tokens)
    return current and ("Ctx " .. current) or snapshot.waiting
  end
  if id == "used_tokens" and usage then
    local counts = pricing.usage_counts(usage)
    if counts.input == nil and counts.output == nil then return nil end
    return "↑" .. (M.compact_number(counts.input) or "—") .. " ↓" .. (M.compact_number(counts.output) or "—")
  end
  if id == "cache_rate" and usage then
    local rate = pricing.cache_rate(usage)
    return rate and ("Cache " .. compact_trim(string.format("%.1f", rate)) .. "%") or "Cache —"
  end
  if id == "cost" and usage then
    return pricing.cost_text(pricing.estimate(snapshot.model, usage, options.pricing))
  end
  if id == "context_used" and usage then
    local remaining = to_number(usage.context_remaining_percent)
    if remaining ~= nil then
      return "Ctx " .. tostring(math.max(0, math.min(100, 100 - math.floor(remaining + 0.5)))) .. "% used"
    end
  end
  if id == "context_window" and usage then
    local value = M.compact_number(usage.context_window)
    return value and (value .. " window") or nil
  end
  if id == "input_tokens" and usage then
    local value = M.compact_number(pricing.usage_counts(usage).input)
    return value and ("↑" .. value) or nil
  end
  if id == "cached_tokens" and usage then
    local value = M.compact_number(usage.cached)
    return value and value ~= "0" and (value .. " cached") or nil
  end
  if id == "output_tokens" and usage then
    local value = M.compact_number(usage.output)
    return value and ("↓" .. value) or nil
  end
  if id == "reasoning_tokens" and usage then
    local value = M.compact_number(usage.reasoning)
    return value and value ~= "0" and (value .. " reasoning") or nil
  end
  return nil
end

local function remove_kind(segments, kind)
  local result = {}
  for _, segment in ipairs(segments) do
    if segment.kind ~= kind then
      table.insert(result, segment)
    end
  end
  return result
end

local function fit_segments(segments, columns, powerline)
  local result = segments
  for _, kind in ipairs(DROP_ORDER) do
    if plan_width(result, powerline) <= columns then
      break
    end
    result = remove_kind(result, kind)
  end
  return result
end

local function with_version(segments, options)
  local theme = type(options.theme) == "table" and options.theme or {}
  table.insert(segments, { kind = "version", text = version.text,
    bg = theme.bg or "#11151a", fg = theme.dim or theme.fg or "#82909f" })
  return segments
end

function M.build_render_plan(options, snapshot, columns)
  options = type(options) == "table" and options or {}
  snapshot = type(snapshot) == "table" and snapshot or {}
  local cols = math.max(1, math.floor(to_number(columns) or 120))
  local layout = M.render_layout(cols)
  local render = type(options.render) == "table" and options.render or {}
  local disabled = list_set(render.disabled_segments)
  local layout_hidden = {}
  if layout == "tiny" then
    layout_hidden = list_set({
      "reasoning", "provider", "personality", "service_tier", "cwd", "project", "git", "permissions",
      "approval", "context_used", "context_window", "used_tokens", "cache_rate", "cost", "input_tokens", "cached_tokens",
      "output_tokens", "reasoning_tokens", "thread_id", "task_progress", "codex_version",
    })
  elseif layout == "narrow" then
    layout_hidden = list_set({
      "provider", "personality", "service_tier", "project", "permissions", "approval", "context_used",
      "context_window", "input_tokens", "cached_tokens", "output_tokens", "reasoning_tokens", "thread_id",
      "task_progress", "codex_version", "cache_rate", "cost",
    })
  end

  local segments = {}
  for _, id in ipairs(ordered_ids(options)) do
    if not disabled[id] and not layout_hidden[id] then
      local text = segment_text(id, options, snapshot, layout)
      if text and tostring(text) ~= "" then
        local bg, fg = theme_for(options, id)
        table.insert(segments, { kind = id, text = tostring(text), bg = bg, fg = fg })
      end
    end
  end

  local powerline = render.powerline ~= false
  local rows = math.max(1, math.min(2, math.floor(to_number(options.bottom_pane and options.bottom_pane.rows) or 1)))
  if rows == 1 then
    return {
      layout = layout,
      columns = cols,
      lines = { fit_segments(with_version(segments, options), cols, powerline) },
    }
  end

  local metadata = {}
  local usage = {}
  for _, segment in ipairs(segments) do
    if list_set({
      "context", "context_used", "context_window", "used_tokens", "cache_rate", "cost", "input_tokens", "cached_tokens",
      "output_tokens", "reasoning_tokens",
    })[segment.kind] then
      table.insert(usage, segment)
    else
      table.insert(metadata, segment)
    end
  end
  metadata = fit_segments(with_version(metadata, options), cols, powerline)
  usage = fit_segments(usage, cols, powerline)
  return {
    layout = layout,
    columns = cols,
    lines = { metadata, usage },
  }
end

M.BASELINE_TOKENS = BASELINE_TOKENS
M.DEFAULT_SEGMENT_ORDER = DEFAULT_SEGMENT_ORDER
M.LEGACY_SEGMENT_ORDER = LEGACY_SEGMENT_ORDER

return M

File: packages/python/src/wezterm_codex_status_line/render.py
from __future__ import annotations

from . import __version__
from .pricing import cache_rate, cost_text, estimate_cost, usage_counts

from copy import deepcopy
from math import floor, isfinite
import re
from typing import Any


SEGMENT_IDS = [
    "label",
    "model",
    "reasoning",
    "activity",
    "provider",
    "personality",
    "service_tier",
    "cwd",
    "project",
    "git",
    "permissions",
    "approval",
    "context",
    "context_used",
    "context_window",
    "used_tokens",
    "cache_rate",
    "cost",
    "input_tokens",
    "cached_tokens",
    "output_tokens",
    "reasoning_tokens",
    "thread_id",
    "task_progress",
    "codex_version",
    "icon",
]
LEGACY_SEGMENT_IDS = [
    "label",
    "model",
    "reasoning",
    "provider",
    "cwd",
    "git",
    "context",
    "used_tokens",
]
DROP_ORDER = [
    "provider",
    "personality",
    "service_tier",
    "codex_version",
    "thread_id",
    "approval",
    "permissions",
    "cached_tokens",
    "reasoning_tokens",
    "input_tokens",
    "output_tokens",
    "context_window",
    "context_used",
    "task_progress",
    "project",
    "cwd",
    "cache_rate",
    "cost",
    "used_tokens",
    "git",
    "reasoning",
    "model",
]
USAGE_SEGMENT_IDS = {
    "context",
    "context_used",
    "context_window",
    "used_tokens",
    "cache_rate",
    "cost",
    "input_tokens",
    "cached_tokens",
    "output_tokens",
    "reasoning_tokens",
}
DEFAULT_ACTIVITY_LABELS = {
    "plan": "PLAN",
    "review": "REVIEW",
    "goal_active": "GOAL",
    "goal_paused": "GOAL PAUSED",
    "goal_blocked": "GOAL BLOCKED",
    "goal_usage_limited": "GOAL LIMITED",
    "goal_budget_limited": "GOAL BUDGET",
    "goal_complete": "GOAL DONE",
}


def _round_half_up(value: float) -> int:
    return floor(value + 0.5)


def compact_number(value: float | int | None) -> str | None:
    if value is None:
        return None
    number = float(value)
    if not isfinite(number):
        return None
    absolute = abs(number)
    sign = "-" if number < 0 else ""
    if absolute < 1000:
        return f"{sign}{_round_half_up(absolute)}"

    units = ((1e12, "T"), (1e9, "B"), (1e6, "M"), (1e3, "K"))
    for index, (divisor, suffix) in enumerate(units):
        if absolute < divisor:
            continue
        scaled = absolute / divisor
        digits = 2 if scaled < 10 else 1 if scaled < 100 else 0
        formatted = f"{scaled:.{digits}f}"
        if float(formatted) >= 1000 and index > 0:
            divisor, suffix = units[index - 1]
            scaled = absolute / divisor
            digits = 2 if scaled < 10 else 1 if scaled < 100 else 0
            formatted = f"{scaled:.{digits}f}"
        if "." in formatted:
            formatted = formatted.rstrip("0").rstrip(".")
        return f"{sign}{formatted}{suffix}"
    return f"{sign}{_round_half_up(absolute)}"


def context_bar(percent: float, cells: int) -> str:
    safe = max(0, min(100, percent))
    width = max(1, int(cells))
    filled = max(0, min(width, _round_half_up((safe / 100) * width)))
    return "█" * filled + "░" * (width - filled)


def render_layout(columns: int) -> str:
    if columns < 60:
        return "tiny"
    if columns < 90:
        return "narrow"
    if columns < 120:
        return "medium"
    return "wide"


def _activity_text(config: dict[str, Any], state: dict[str, Any]) -> str | None:
    activity = state.get("activity") or {}
    labels = config["options"].get("activity", {}).get("labels", {})
    if activity.get("review"):
        return labels.get("review") or DEFAULT_ACTIVITY_LABELS["review"]
    if activity.get("mode") == "plan":
        return labels.get("plan") or DEFAULT_ACTIVITY_LABELS["plan"]
    goal = activity.get("goal") or {}
    status = goal.get("status")
    key_by_status = {
        "active": "goal_active",
        "paused": "goal_paused",
        "blocked": "goal_blocked",
        "usageLimited": "goal_usage_limited",
        "budgetLimited": "goal_budget_limited",
        "complete": "goal_complete",
    }
    key = key_by_status.get(status)
    return (labels.get(key) or DEFAULT_ACTIVITY_LABELS[key]) if key else None


MODEL_ICONS = {"astra": "✦", "sol": "☀", "terra": "⊕", "luna": "☾"}


def _model_text(model: str | None, display: str = "name") -> str | None:
    if not model or display not in {"icon_name", "icon"}:
        return model
    for token in re.findall(r"[a-z0-9]+", model.lower()):
        if icon := MODEL_ICONS.get(token):
            return icon if display == "icon" else f"{icon} {model}"
    return model


def _segment_text(
    segment_id: str,
    config: dict[str, Any],
    state: dict[str, Any],
    layout: str,
) -> str | None:
    options = config["options"]
    if segment_id == "label":
        return state.get("label") or options.get("label") or "CODEX"
    if segment_id == "model":
        return _model_text(state.get("model"), options["render"].get("model_display", "name"))
    if segment_id in {"reasoning", "cwd", "project", "git", "permissions", "approval"}:
        return state.get(segment_id)
    if segment_id == "activity":
        return _activity_text(config, state)
    if segment_id == "provider":
        provider = state.get("provider")
        return f"p:{provider}" if provider and layout == "wide" else provider
    if segment_id == "personality":
        return f"persona:{state['personality']}" if state.get("personality") else None
    if segment_id == "service_tier":
        return f"tier:{state['service_tier']}" if state.get("service_tier") else None
    if segment_id == "thread_id":
        return f"id:{str(state['thread_id'])[:8]}" if state.get("thread_id") else None
    if segment_id == "task_progress":
        progress = state.get("task_progress") or {}
        completed = progress.get("completed")
        total = progress.get("total")
        return f"Tasks {floor(completed)}/{floor(total)}" if completed is not None and total else None
    if segment_id == "codex_version":
        version = str(state.get("codex_version") or "").removeprefix("v")
        return f"v{version}" if version else None
    if segment_id == "icon":
        return options.get("icon", {}).get("text") or ""

    usage = state.get("usage")
    if segment_id == "used_tokens" and usage is not None:
        counts = usage_counts(usage)
        if counts["input"] is None and counts["output"] is None:
            return None
        return f"↑{compact_number(counts['input']) or '—'} ↓{compact_number(counts['output']) or '—'}"
    if segment_id == "cache_rate" and usage is not None:
        rate = cache_rate(usage)
        return "Cache —" if rate is None else f"Cache {rate:g}%"
    if segment_id == "cost" and usage is not None:
        return cost_text(estimate_cost(state.get("model"), usage, options.get("pricing")))
    if segment_id == "context":
        if not usage:
            return state.get("waiting")
        percent = usage.get("context_remaining_percent")
        if percent is not None:
            rounded = max(0, min(100, _round_half_up(float(percent))))
            if layout in {"wide", "medium"}:
                cells = 10 if layout == "wide" else 8
                return f"Ctx {context_bar(rounded, cells)} {rounded}% left"
            return f"Ctx {rounded}% left" if layout == "narrow" else f"Ctx {rounded}%"
        current = compact_number(usage.get("context_tokens"))
        return f"Ctx {current}" if current else state.get("waiting")
    if not usage:
        return None
    if segment_id == "context_used" and usage.get("context_remaining_percent") is not None:
        remaining = max(0, min(100, _round_half_up(float(usage["context_remaining_percent"]))))
        return f"Ctx {100 - remaining}% used"
    value_by_segment = {
        "context_window": (usage.get("context_window"), " window", False),
        "input_tokens": (usage_counts(usage)["input"], "", False),
        "cached_tokens": (usage.get("cached"), " cached", True),
        "output_tokens": (usage.get("output"), "", False),
        "reasoning_tokens": (usage.get("reasoning"), " reasoning", True),
    }
    if segment_id not in value_by_segment:
        return None
    value, suffix, omit_zero = value_by_segment[segment_id]
    compact = compact_number(value)
    prefix = {"input_tokens": "↑", "output_tokens": "↓"}.get(segment_id, "")
    return f"{prefix}{compact}{suffix}" if compact and (not omit_zero or compact != "0") else None


def _plan_width(segments: list[dict[str, str]], powerline: bool) -> int:
    return sum(
        len(segment["text"])
        + (2 if powerline else 0)
        + ((1 if powerline else 2) if index else 0)
        for index, segment in enumerate(segments)
    )


def _fit(segments: list[dict[str, str]], columns: int, powerline: bool) -> list[dict[str, str]]:
    result = deepcopy(segments)
    for segment_id in DROP_ORDER:
        if _plan_width(result, powerline) <= columns:
            break
        result = [segment for segment in result if segment["kind"] != segment_id]
    return result


def _theme_for(options: dict[str, Any], segment_id: str) -> dict[str, str]:
    theme = options["theme"]
    segment_theme = theme.get("segments", {}).get(segment_id) or {}
    return {
        "bg": segment_theme.get("bg") or theme["bg"],
        "fg": segment_theme.get("fg") or theme["fg"],
    }


def _ordered_ids(configured: list[str]) -> list[str]:
    order = list(dict.fromkeys(configured))
    appendable = ["activity", "icon"] if all(item in LEGACY_SEGMENT_IDS for item in order) else SEGMENT_IDS
    expanded = list(dict.fromkeys([*order, *appendable]))
    return [item for item in expanded if item != "icon"] + [item for item in expanded if item == "icon"]


def build_render_plan(config: dict[str, Any], state: dict[str, Any], columns: int) -> dict[str, Any]:
    columns = max(1, int(columns or 120))
    layout = render_layout(columns)
    options = config["options"]
    render = options["render"]
    disabled = set(render.get("disabled_segments", []))
    if layout == "tiny":
        hidden = {
            "reasoning", "provider", "personality", "service_tier", "cwd", "project", "git",
            "permissions", "approval", "context_used", "context_window", "used_tokens", "cache_rate", "cost",
            "input_tokens", "cached_tokens", "output_tokens", "reasoning_tokens", "thread_id",
            "task_progress", "codex_version",
        }
    elif layout == "narrow":
        hidden = {
            "provider", "personality", "service_tier", "project", "permissions", "approval",
            "context_used", "context_window", "input_tokens", "cached_tokens", "output_tokens",
            "reasoning_tokens", "thread_id", "task_progress", "codex_version", "cache_rate", "cost",
        }
    else:
        hidden = set()

    configured = render.get("segment_order") or SEGMENT_IDS
    segments: list[dict[str, str]] = []
    for segment_id in _ordered_ids(configured):
        if segment_id in disabled or segment_id in hidden:
            continue
        text = _segment_text(segment_id, config, state, layout)
        if not text:
            continue
        segments.append({"kind": segment_id, "text": text, **_theme_for(options, segment_id)})

    version = {
        "kind": "version", "text": f"v{__version__}",
        "bg": options["theme"]["bg"], "fg": options["theme"].get("dim") or options["theme"]["fg"],
    }
    powerline = render.get("powerline", True)
    if options["bottom_pane"]["rows"] == 1:
        lines = [_fit([*segments, version], columns, powerline)]
    else:
        metadata = [segment for segment in segments if segment["kind"] not in USAGE_SEGMENT_IDS]
        lines = [
            _fit([*metadata, version], columns, powerline),
            _fit([segment for segment in segments if segment["kind"] in USAGE_SEGMENT_IDS], columns, powerline),
        ]
    return {"layout": layout, "columns": columns, "lines": lines}
